### PR TITLE
pseudo-randomness generator

### PR DESCRIPTION
**Pull request by andrei-diaconescu**

### CODE REVIEW

File: arwen/contexts/managedType.go [hunk 1]
```diff
@@ -56,6 +58,19 @@ func NewManagedTypesContext(host arwen.VMHost) (*managedTypesContext, error) {
...
 
+func initRandomizer(host arwen.VMHost) *math.SeedRandReader {
+	randomizer, _ := math.NewSeedRandReader(append(host.Blockchain().CurrentRandomSeed(), host.Runtime().GetCurrentTxHash()...))
+
+	return randomizer
+}
+
+func (context *managedTypesContext) GetRandReader(ctx unsafe.Pointer) *math.SeedRandReader {
+	if context.Reader == nil {
+		context.Reader = initRandomizer(arwen.GetVMHost(ctx))
```
Comment (sasurobert) [inline arwen/contexts/managedType.go:69]: you do not need to do GetVMHost - you already have access as context.host

Comment (sasurobert) [reply arwen/contexts/managedType.go:69]: So you do not need the unsafe.Pointer as input.

File: arwen/contexts/managedType.go [hunk 2]
```diff
@@ -56,6 +58,19 @@ func NewManagedTypesContext(host arwen.VMHost) (*managedTypesContext, error) {
 	return context, nil
 }
 
+func initRandomizer(host arwen.VMHost) *math.SeedRandReader {
+	randomizer, _ := math.NewSeedRandReader(append(host.Blockchain().CurrentRandomSeed(), host.Runtime().GetCurrentTxHash()...))
+
+	return randomizer
+}
+
+func (context *managedTypesContext) GetRandReader(ctx unsafe.Pointer) *math.SeedRandReader {
```
Comment (sasurobert) [inline arwen/contexts/managedType.go:67]: add comment to public function

File: math/randomnessGenerator.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	"crypto/sha256"
+	"encoding/binary"
+	"math/rand"
+)
+
+type SeedRandReader struct {
+	rand *rand.Rand
+}
+
+func NewSeedRandReader(seed []byte) (*SeedRandReader, error) {
```
Comment (sasurobert) [inline math/randomnessGenerator.go:13]: add comment to public function

File: arwen/contexts/managedType.go [hunk 4]
```diff
@@ -56,6 +58,19 @@ func NewManagedTypesContext(host arwen.VMHost) (*managedTypesContext, error) {
...
+	if context.Reader == nil {
```
Comment (sasurobert) [inline arwen/contexts/managedType.go:68]: do check.IfNil - instead of simple nil check here. make reader implement IsInterfaceNil. Make Reader private as well.

File: arwen/contexts/managedType.go [hunk 5]
```diff
@@ -56,6 +58,19 @@ func NewManagedTypesContext(host arwen.VMHost) (*managedTypesContext, error) {
 	return context, nil
 }
 
+func initRandomizer(host arwen.VMHost) *math.SeedRandReader {
+	randomizer, _ := math.NewSeedRandReader(append(host.Blockchain().CurrentRandomSeed(), host.Runtime().GetCurrentTxHash()...))
```
Comment (sasurobert) [inline arwen/contexts/managedType.go:62]: treat error here. or make NewSeedRandReader to not return error.

File: math/randomnessGenerator.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	if len(seed) == 0 {
```
Comment (sasurobert) [inline math/randomnessGenerator.go:14]: no need for this check.

File: math/randomnessGenerator.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	rand := rand.New(source)
+
+	return &SeedRandReader{
+		rand: rand,
+	}, nil
+}
+
+// Read will read upto len(p) bytes. It will rotate the existing byte buffer (seed) until it will fill up the provided
+// p buffer
+func (srr *SeedRandReader) Read(p []byte) (n int, err error) {
```
Comment (sasurobert) [inline math/randomnessGenerator.go:31]: implement IsInterfaceNil

File: arwen/contexts/managedType.go (hunk 2 above)
Comment (sasurobert) [reply arwen/contexts/managedType.go:67]: and use and Interface as return here. and keep the context.Reader as interface as well.

File: math/randomnessGenerator.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+		return nil, ErrSeedLengthIsZero
+	}
+
+	seedHash := sha256.Sum256(seed)
+	seedNumber := binary.BigEndian.Uint64(seedHash[:])
+
+	source := rand.NewSource(int64(seedNumber))
+	rand := rand.New(source)
+
+	return &SeedRandReader{
```
Comment (sasurobert) [inline math/randomnessGenerator.go:24]: unit tests here and in all new implementation please

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/contexts/managedType.go (hunk 5 above)
Comment (andrei-diaconescu) [reply arwen/contexts/managedType.go:62]: it cannot return error here, that's why error is not treated. len(seed) > 0 in this specific use case.

File: arwen/contexts/managedType.go [hunk 9]
```diff
@@ -4,6 +4,7 @@ import (
 	"crypto/elliptic"
 	basicMath "math"
 	"math/big"
+	"unsafe"
```
Comment (iulianpascalau) [inline arwen/contexts/managedType.go:7]: do not use unsafe

File: arwen/contexts/managedType.go [hunk 10]
```diff
@@ -33,6 +34,7 @@ type managedTypesContext struct {
 	host               arwen.VMHost
 	managedTypesValues managedTypesState
 	managedTypesStack  []managedTypesState
+	Reader             *math.SeedRandReader
```
Comment (iulianpascalau) [inline arwen/contexts/managedType.go:37]: use interface

File: math/randomnessGenerator.go (hunk 7 above)
Comment (iulianpascalau) [reply math/randomnessGenerator.go:31]: is this going to be called in a concurrent manner?

File: math/randomnessGenerator.go [hunk 11]
```diff
@@ -0,0 +1,37 @@
+package math
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"math/rand"
+)
+
+type SeedRandReader struct {
```
Comment (iulianpascalau) [inline math/randomnessGenerator.go:9]: make unexported & add tests.
Please do add tests that check that for a predefined string we always get the same n "random" bytes. In this way we have the certainty that we won't add backwards compatibility bugs. (the predefined strings should have some comments stating that those strings should not change)

File: math/randomnessGenerator.go [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+	source := rand.NewSource(int64(seedNumber))
+	rand := rand.New(source)
+
+	return &SeedRandReader{
+		rand: rand,
+	}, nil
+}
+
+// Read will read upto len(p) bytes. It will rotate the existing byte buffer (seed) until it will fill up the provided
```
Comment (iulianpascalau) [inline math/randomnessGenerator.go:29]: Are we sure the comment does what it states? Maybe update the comment? (I remember this comment from the first implementation done by me)

File: math/errors.go [hunk 13]
```diff
@@ -12,3 +12,6 @@ var ErrSubtractionUnderflow = errors.New("subtraction underflow")
 
 // ErrMultiplicationOverflow is raised when there is an overflow because of the multiplication of two numbers
 var ErrMultiplicationOverflow = errors.New("multiplication overflow")
+
+// ErrSeedLengthIsZero is raised when length of given seed is 0
+var ErrSeedLengthIsZero = errors.New("seed length cannot be 0")
```
Comment (iulianpascalau) [inline math/errors.go:17]: this should be removed when resolving @sasurobert comment

File: arwen/contexts/managedType.go [hunk 14]
```diff
@@ -56,6 +58,19 @@ func NewManagedTypesContext(host arwen.VMHost) (*managedTypesContext, error) {
 	return context, nil
 }
 
+func initRandomizer(host arwen.VMHost) *math.SeedRandReader {
```
Comment (iulianpascalau) [inline arwen/contexts/managedType.go:61]: + add a unit test. Same here: hardcoded values that will always produce the same outcome. Backwards compatibility and deterministic results are important.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: math/randomnessGenerator.go (hunk 7 above)
Comment (sasurobert) [reply math/randomnessGenerator.go:31]: no. never.

File: arwen/contexts/managedType_test.go [hunk 15]
```diff
@@ -29,6 +31,49 @@ func TestNewManagedTypes(t *testing.T) {
...
+	prg := managedTypesContext.GetRandReader()
+	a := make([]byte, 100)
+	prg.Read(a)
+	b := make([]byte, 100)
+	for i := 0; i < 1000; i++ {
+		prg.Read(b)
+		require.NotEqual(t, a, b)
+		fmt.Println(hex.EncodeToString(a))
+		fmt.Println(hex.EncodeToString(b))
+		copy(a, b)
```
Comment (iulianpascalau) [inline arwen/contexts/managedType_test.go:73]: why this?

File: math/interface.go [hunk 16]
```diff
@@ -0,0 +1,6 @@
+package math
+
+type RandomnessGenerator interface {
```
Comment (iulianpascalau) [inline math/interface.go:3]: Missing comment

File: math/randomnessGenerator_test.go [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+	require.True(t, firstRandomizer.IsInterfaceNil())
+	firstRandomizer = NewSeedRandReader(firstSeed)
+	require.False(t, firstRandomizer.IsInterfaceNil())
+
+	secondStr := "the usage of the same seed must generate 2 equal randomizers"
+	secondSeed := []byte(secondStr)
+	var secondRandomizer *seedRandReader
+	require.True(t, secondRandomizer.IsInterfaceNil())
+	secondRandomizer = NewSeedRandReader(secondSeed)
+	require.Equal(t, firstRandomizer, secondRandomizer)
```
Comment (iulianpascalau) [inline math/randomnessGenerator_test.go:22]: why this?

File: math/randomnessGenerator_test.go [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+	var thirdRandomizer *seedRandReader
+	require.True(t, thirdRandomizer.IsInterfaceNil())
+	thirdRandomizer = NewSeedRandReader(thirdSeed)
+	require.NotEqual(t, firstRandomizer, thirdRandomizer)
+	require.NotEqual(t, secondRandomizer, thirdRandomizer)
+	require.False(t, thirdRandomizer.IsInterfaceNil())
+
+	a := make([]byte, 100)
+	firstRandomizer.Read(a)
+	require.NotEqual(t, firstRandomizer, secondRandomizer)
```
Comment (iulianpascalau) [inline math/randomnessGenerator_test.go:36]: can be removed

File: math/randomnessGenerator_test.go [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+	b := make([]byte, 100)
+	secondRandomizer.Read(b)
+	require.Equal(t, a, b)
+	require.Equal(t, firstRandomizer, secondRandomizer)
```
Comment (iulianpascalau) [inline math/randomnessGenerator_test.go:40]: can be removed

File: math/randomnessGenerator.go [hunk 20]
```diff
@@ -6,32 +6,29 @@ import (
...
-
-// Read will read upto len(p) bytes. It will rotate the existing byte buffer (seed) until it will fill up the provided
-// p buffer
-func (srr *SeedRandReader) Read(p []byte) (n int, err error) {
-	if len(p) == 0 {
-		return 0, ErrSeedLengthIsZero
 	}
+}
 
+// Read generates len(p) random bytes and writes them into p. It always returns len(p) and a nil error.
```
Comment (iulianpascalau) [inline math/randomnessGenerator.go:26]: are we sure it always returns nil error? What about if I provide a nil buffer? Please add unit tests for nil buffer & empty buffer. Check what it returns.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: math/randomnessGenerator.go (hunk 7 above)
Comment (iulianpascalau) [reply math/randomnessGenerator.go]: great :+1:

File: arwen/contexts/managedType_test.go (hunk 15 above)
Comment (iulianpascalau) [reply arwen/contexts/managedType_test.go:73]: for the compare, figured out later. All good :+1:

File: math/randomnessGenerator.go (hunk 20 above)
Comment (iulianpascalau) [reply math/randomnessGenerator.go:26]: Can add a separate test called "...._BackwardsCompatibilityCheck...."

File: math/randomnessGenerator_test.go [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+	c := make([]byte, 100)
+	thirdRandomizer.Read(c)
+	require.NotEqual(t, a, c)
+}
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply math/randomnessGenerator_test.go:44]: Can add a separate test called "...._BackwardsCompatibilityCheck...."

File: arwen/contexts/managedType.go [hunk 22]
```diff
@@ -33,6 +35,7 @@ type managedTypesContext struct {
 	host               arwen.VMHost
 	managedTypesValues managedTypesState
 	managedTypesStack  []managedTypesState
+	srr                math.RandomnessGenerator
```
Comment (sasurobert) [inline arwen/contexts/managedType.go:38]: change srr to randomnessGenerator

File: math/randomnessGenerator_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply math/randomnessGenerator_test.go:44]: There is no need for backward compatibility check - as this is not an active function on mainnet.

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by sasurobert — APPROVED** [review]

File: math/randomnessGenerator_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply math/randomnessGenerator_test.go:62]: Will be at some point. I don't want to debug a one liner future bug.

**Review by iulianpascalau — APPROVED** [review]
